### PR TITLE
Calculate offsets for weakreflist and dict in PyCell.

### PR DESCRIPTION
**Pull request by sebpuetz**
As demonstrated by #1022, depending on the order of structmembers to calculate offsets in bytes is fragile. ~~This PR adds a dependency to `memoffset` which provides an interface to find offsets into structs in a more robust manner.~~

~~Caveat: This macro is technically unsound, see https://github.com/Gilnaa/memoffset/issues/24, so I'm marking this as a draft for now.~~

This PR moves the calcualtion of class dict and weakreflist offsets to `PyCell` rather than `initialize_type_object`.

-----------------
Thank you for contributing to pyo3!

Here are some things you should check for submitting your pull request:

 - Run `cargo fmt` (This is checked by travis ci)
 - Run `cargo clippy` and check there are no hard errors (There are a bunch of existing warnings; This is also checked by travis)
 - If applicable, add an entry in the changelog.
 - If applicable, add documentation to all new items and extend the guide.
 - If applicable, add tests for all new or fixed functions
 - If you changed any python code, run `black .`. You can install black with `pip install black`)

You might want to run `tox` (`pip install tox`) locally to check compatibility with all supported python versions. If you're using linux or mac you might find the Makefile helpful for testing.

### CODE REVIEW

**Comment by sebpuetz** [comment]
An alternative implementation that makes maintenance easier without pulling in unsoundness:

~~~Rust
#[repr(C)]
pub struct PyCell<T: PyClass> {
    inner: PyCellInner<T>,
    thread_checker: T::ThreadChecker,
    // DO NOT CHANGE THE ORDER OF THESE FIELDS WITHOUT CHANGING PyCell::dict_offset()
    // AND PyCell::weakref_offset()
    dict: T::Dict,
    weakref: T::WeakRef,
}

impl<T: PyClass> PyCell<T> {
    /// Get the offset of the dictionary from the start of the struct in bytes.
    pub(crate) fn dict_offset() -> usize {
        mem::size_of::<Self>() - mem::size_of::<T::Dict>() - mem::size_of::<T::WeakRef>()
    }

    /// Get the offset of the weakref list from the start of the struct in bytes.
    pub(crate) fn weakref_offset() -> usize {
        mem::size_of::<Self>() - mem::size_of::<T::WeakRef>()
    }
}
~~~

**Comment by sebpuetz** [comment]
@kngwyu

File: src/pyclass.rs [hunk 1]
```diff
@@ -137,18 +137,14 @@ where
     // type size
     type_object.tp_basicsize = std::mem::size_of::<T::Layout>() as ffi::Py_ssize_t;
 
-    let mut offset = type_object.tp_basicsize;
-
     // __dict__ support
-    if let Some(dict_offset) = T::Dict::OFFSET {
-        offset += dict_offset as ffi::Py_ssize_t;
-        type_object.tp_dictoffset = offset;
+    if !T::Dict::IS_DUMMY {
```
Comment (kngwyu) [inline src/pyclass.rs:141]: How about moving this if statement to the `PyCell`?
I don't like that `dict_offset` can return a meaningless value even if it doesn't have a dict.

**Comment by kngwyu** [comment]
Thanks!

File: src/pyclass.rs (hunk 1 above)
Comment (sebpuetz) [reply src/pyclass.rs:141]: Good catch, I `PyCell::*_offset` now returns `Option<usize>`

**Comment by sebpuetz** [comment]
Do you want to add a commit with the test case you wrote for #1061?

**Comment by kngwyu** [comment]
> Do you want to add a commit with the test case you wrote for #1061?

Yeah, could you please copy-paste it to this branch?

**Comment by sebpuetz** [comment]
Should be good to go if CI passes

**Comment by kngwyu** [comment]
Thanks!

**Pull request merged**